fs/operations/operations.go

```diff
@@ -830,59 +830,144 @@ func ListLong(ctx context.Context, f fs.Fs, w io.Writer) error {
 // excludes
 //
 // Lists in parallel which may get them out of order
-func Md5sum(ctx context.Context, f fs.Fs, w io.Writer) error {
-	return HashLister(ctx, hash.MD5, f, w)
-}
+// func Md5sum(ctx context.Context, f fs.Fs, w io.Writer) error {
+// 	return HashLister(ctx, hash.MD5, false, false, f, w)
+// }
 
 // Sha1sum list the Fs to the supplied writer
 //
 // Obeys includes and excludes
 //
 // Lists in parallel which may get them out of order
-func Sha1sum(ctx context.Context, f fs.Fs, w io.Writer) error {
-	return HashLister(ctx, hash.SHA1, f, w)
-}
+// func Sha1sum(ctx context.Context, f fs.Fs, w io.Writer) error {
+// 	return HashLister(ctx, hash.SHA1, false, false, f, w)
+// }
 
 // hashSum returns the human readable hash for ht passed in.  This may
 // be UNSUPPORTED or ERROR. If it isn't returning a valid hash it will
 // return an error.
-func hashSum(ctx context.Context, ht hash.Type, o fs.Object) (string, error) {
+func hashSum(ctx context.Context, ht hash.Type, downloadFlag bool, o fs.Object) (string, error) {
+	var sum string
 	var err error
-	tr := accounting.Stats(ctx).NewCheckingTransfer(o)
-	defer func() {
-		tr.Done(ctx, err)
-	}()
-	sum, err := o.Hash(ctx, ht)
-	if err == hash.ErrUnsupported {
-		sum = "UNSUPPORTED"
-	} else if err != nil {
-		fs.Debugf(o, "Failed to read %v: %v", ht, err)
-		sum = "ERROR"
+
+	// If downloadFlag is true, download and hash the file.
+	// If downloadFlag is false, call o.Hash asking the backend for the hash
+	if downloadFlag {
+		
+		// Setup accounting as NewTransfer.  Same approach used in Cat
+		tr := accounting.Stats(ctx).NewTransfer(o)
+		defer func() {
+			tr.Done(ctx, err)
+		}()
+
+		// Setup download options.  Similar approach to Cat.
+		// We do not have a start / end, so we only use fs.Config.DownloadHeaders.
+		// Unsure if copying to a new OpenOption object is required.
+		// Copy also has a very similar setup but includes hashOption
+		var options []fs.OpenOption
+		for _, option := range fs.Config.DownloadHeaders {
+			options = append(options, option)
+		}
+
+		// Using NewReOpen.  Error handling seems to be done different everwhere.  This version is from Cat
```

Comment:
Confirm usage of NewReOpen and how error handling should be done for this

Reply:
When you get an error you should run fs.CountError() on it and log it at some level, probably ERROR

Normally rclone wouldn't do this if you are returning the error, but the callers of hashSum don't log the error or return it to their callers so it is necessary.

I think the error handling here needs a re-write, I would not call `fs.CountError()` or `fs.Errorf` but I would return the error wrapped `return "ERROR", errors.Wrap(err, "failed to open")`

Then the caller of `hashLister` should log these errors with `fs.CountError()` and `fs.Errorf` and return the first one as its own ERROR.

Reply:
> For example, I see copy doing CountError and Errorf in a similar use case.

Because it hasn't been tidyed. It might be if we return to that code later. You touch your part, you look at it, you decide tidying is good. You tidy. You don't touch `Copy`. It stays as is. Step by step.

> If the caller of hashLister needs to log the errors, does this mean we should log them all the way back in the commands?

No. Why log repeatedly?